File: day7/【学长带练】观光公交.py
n, m, k = map(int, input().split())
D = [0] + [int(x) for x in input().split()]
T, A, B = [], [], []

last = [0] * n
people = [0] * n
arrive = [0] * n


def update_arrive():
    for i in range(1, n):
        arrive[i] = D[i] + max(arrive[i - 1], last[i - 1])


def count_time():
    cnt = 0
    for i in range(m):
        b = B[i]
        cnt += arrive[b] - T[i]
    return cnt


def calc_reduce():
    reduce = people.copy()
    
    for i in range(n - 2, 0, -1):
        if arrive[i] > last[i]:
            reduce[i] += reduce[i + 1]
    return reduce


def most_reduce():
    reduce = calc_reduce()

    p = 0
    for i in range(1, n):
        if D[i] and reduce[i] > reduce[p]:
            p = i
    return p


if __name__ == '__main__':
    for _ in range(m):
        t, a, b = map(int, input().split())
        T.append(t)
        A.append(a - 1)
        B.append(b - 1)

    for t, a, b in zip(T, A, B):
        last[a] = max(last[a], t)

        people[b] += 1

    update_arrive()
    for _ in range(k):
        D[most_reduce()] -= 1

        update_arrive()

    print(count_time())

File: day7/【学长带练】观光公交.cpp
#include <cstdio>
#include <cstring>
#include <iostream>
#include <algorithm>

using namespace std;

const int N = 1010, M = 10010;

int n, m, k;
int d[N];
int t[M], a[M], b[M];
int last[N], sum[N];
int tm[N], reduce[N];

int main()
{
    scanf("%d%d%d", &n, &m, &k);

    for (int i = 1; i < n; i++) scanf("%d", &d[i]);
    for (int i = 0; i < m; i++)
    {
        scanf("%d%d%d", &t[i], &a[i], &b[i]);
        last[a[i]] = max(last[a[i]], t[i]);
        sum[b[i]] ++;
    }

    for (int i = 1; i <= n; i++) tm[i] = max(tm[i - 1], last[i - 1]) + d[i - 1];

    while (k--)
    {
        for (int i = n; i >= 2; i--)
        {
            reduce[i - 1] = sum[i];
            if (tm[i] > last[i]) reduce[i - 1] += reduce[i];
        }

        int p = 0;
        for (int i = 1; i <= n; i++)
            if (d[i] && reduce[p] < reduce[i])
                p = i;
        d[p] --;

        for (int i = p; i <= n; i++) tm[i] = max(tm[i - 1], last[i - 1]) + d[i - 1];
    }

    int res = 0;
    for (int i = 0; i < m; i++) res += tm[b[i]] - t[i];

    printf("%d\n", res);

    return 0;
}